metagraph/core/dask/resolver.py

```diff
@@ -0,0 +1,185 @@
+        self.class_to_concrete = self._resolver.class_to_concrete.copy()
+        for ct in self._resolver.concrete_types:
+            ph = self._get_placeholder(ct)
+            self.class_to_concrete[ph] = ct
+
+    # Default behavior (unless overridden) is to delegate to the original resolver
+    def __getattr__(self, item):
+        obj = getattr(self._resolver, item)
+        if type(obj) is types.MethodType and type(obj.__self__) is not type:
+            # Replace original resolver with the dask resolver for instance methods
+            return obj.__func__.__get__(self)
+        return obj
+
+    def __dir__(self):
+        return dir(self._resolver)
+
+    def delayed_wrapper(self, klass, concrete_type: Optional[ConcreteType] = None):
+        ct = concrete_type
+        if ct is None:
+            ct = self._resolver.class_to_concrete.get(klass)
+            if ct is None:
+                raise TypeError(
+                    f"{klass.__name__} is not a defined `value_type`. Must provide `concrete_type`."
+                )
+        ph = self._get_placeholder(ct)
+        return DelayedWrapper(klass, ph)
+
+    def _get_placeholder(self, concrete_type):
+        if concrete_type not in self._placeholders:
+            ph = types.new_class(f"{concrete_type.__name__}Placeholder", (Placeholder,))
+            ph.concrete_type = concrete_type
+            self._placeholders[concrete_type] = ph
+        return self._placeholders[concrete_type]
+
+    def _add_translation_plan(self, mst, src, **props):
+        obj = src
+        src_type = mst.src_type
+        for trans, dst_type in zip(mst.translators, mst.dst_types):
+            ph = self._get_placeholder(dst_type)
+            key = (
+                f"translate-{tokenize(ph, trans, obj, props)}",
+                f"{src_type.__name__}->{dst_type.__name__}",
+            )
+            kwargs = {}
+            # Pass in the real resolver if needed by the translator
+            if trans._include_resolver:
+                kwargs["resolver"] = self._resolver
+            if dst_type is mst.final_type:
+                kwargs.update(props)
+            obj = ph.build(key, trans, (obj,), kwargs)
+            src_type = dst_type
+        return obj
+
+    def _add_algorithm_plan(self, algo_plan, *args, **kwargs):
+        args, kwargs = algo_plan.apply_abstract_defaults(
+            self._resolver, algo_plan.algo.abstract_name, *args, **kwargs
+        )
+        sig = algo_plan.algo.__signature__
+        # Walk through arguments and apply required translations
+        bound = sig.bind(*args, **kwargs)
+        bound.apply_defaults()
+        for name, trans in algo_plan.required_translations.items():
+            bound.arguments[name] = self._add_translation_plan(
+                trans, bound.arguments[name]
+            )
+        args, kwargs = bound.args, bound.kwargs
+        # Add resolver if needed by the algorithm
+        if algo_plan.algo._include_resolver:
```

Comment:
I noticed this code isn't covered by the tests, but when I tried writing a test, I discovered that the `include_resolver`  support for algorithms is not working in general.